Networking Demo AppTests/APIClientTests.swift

```diff
@@ -191,19 +190,171 @@ class APIClientTests: XCTestCase {
+        let validGetRequest = GetUsersRequest()
+        do {
+            _ = try await sut.perform(validGetRequest)
+        } catch {
+            XCTFail("Exception thrown while validating response")
+        }
+        
+        await waitForExpectations(timeout: 10, handler: nil)
+    }
+    
+    @available(iOS 15.0.0, *)
+    func testFileDownloadValidationForValidationSuccess() async throws {
+        let responseValidationException = expectation(description: "Expect response validator to be called")
+        let downloadProgressExpectation = expectation(description: "Expect download progress to reach 100%.")
+        
+        let request = TestAsyncDownloadRequest { progress in
+            if progress == 1.0 {
+                downloadProgressExpectation.fulfill()
+            }
+            print("Download progress: \(progress)")
+        }
+        let mockValidator = MockResponseValidator()
+        mockValidator.onValidated = {
+            responseValidationException.fulfill()
+            return
+        }
+        let api = APIClient(apiContext: SimpleAPIContext(), responseValidator: mockValidator)
+        
+        let result: DownloadResponse = try await api.perform(request).responseBody
+        do {
+            let image = UIImage(data: try Data(contentsOf: result.url))
+            XCTAssertNotNil(image)
+            try FileManager.default.removeItem(at: result.url)
+        } catch {
+            XCTFail(error.localizedDescription)
+        }
+        await waitForExpectations(timeout: 10, handler: nil)
+    }
+    
+    @available(iOS 15.0.0, *)
+    func testFileDownloadValidationForValidationFailed() async throws {
+        let responseValidationException = expectation(description: "Expect response validator to be called")
+        let downloadProgressExpectation = expectation(description: "Expect download progress to reach 100%.")
+        
+        let request = TestAsyncDownloadRequest { progress in
+            XCTAssertFalse(Thread.isMainThread, "Shouldn't be on main thread")
+            if progress == 1.0 {
+                downloadProgressExpectation.fulfill()
+            }
+            print("Download progress: \(progress)")
+        }
+        let mockValidator = MockResponseValidator()
+        mockValidator.onValidated = {
+            XCTAssertFalse(Thread.isMainThread, "Shouldn't be on main thread")
+            responseValidationException.fulfill()
+            throw ValidationError.invalidResponse
+        }
+        let api = APIClient(apiContext: SimpleAPIContext(), responseValidator: mockValidator)
+        
+        do {
+            _ = try await api.perform(request).responseBody
+            XCTFail("Expected validation exception to be thrown")
+        } catch {
+            FileManager.default.clearTmpDirectory()
+        }
+        await waitForExpectations(timeout: 10, handler: nil)
+    }
+    
+    // todo remove 
```

Comment:
Is this "TODO" to remove? of TODO "to remove TODO" ? 😆

Reply:
haha good question :D this test needs to be removed, my bad. i was initially planning on testing some threading stuff with this, but turns out there's not much i can actually assert, given the other discussions with Mo. So i've removed that now, thanks